rfcs/45-stageless.md

```diff
@@ -0,0 +1,727 @@
+   2. This can be easily added later, or third-party plugins can create their own abstraction.
+4. When a state transition occurs, the systems in the while-active set are no longer applied immediately.
+   1. Instead, the normal flow of logic continues.
+   2. Users can duplicate critical logic in the while-active and on-enter collections.
+   3. Similarly, arbitrarily long chains of state transitions are no longer be processed in the same schedule pass.
+   4. However, users can add as many copies of the `flush_state` system as they would like, and loop it within an exclusive system.
+
+## Rationale and alternatives
+
+### Doesn't set-configuration allow you to configure systems added by third-party plugins?
+
+Yes. This is a large part of the point.
+
+The difficulties configuring plugins is a source of [significant user pain](https://github.com/bevyengine/bevy/issues/2160).
+Plugin authors cannot know all of the needs of the end user's app, and so some degree of control must be handed back to the end users.
+
+This control is carefully limited though: systems can only be configured if they are part of a public set, and all of the systems under a common set are configured as a group.
+In addition, configuration defined by a private set (or directly on systems) cannot be altered, extended or removed by the app: it cannot be named, and so cannot be altered.
+
+This limitation allows plugin authors to carefully design a public API, ensure critical invariants hold, and make serious internal changes without breaking their dependencies (as long as the public sets and their meanings are relatively stable).
+
+### Why are run criteria shared?
+
+Rather than sharing run criteria within a tick, we could evaluate them once for each system.
+This would have a small performance impact, as in practice, almost all run criteria are extremely simple, and fast to check.
+
+However, this behavior is unintuitive, and can result in desynchronization between members with the same run criteria, potentially leading to serious logic errors.
+
+This design can worked around this by adding duplicate run criteria to systems that require atomic operation, but the opposite workaround is less reliable, involving the creation of dummy resources.
+Atomic groups may improve this level of control in the future.
+
+## What sugar should we use for adding multiple systems at once?
+
+When using convenience methods like `add_systems`, we need to be able to refer to collections of systems.
+This will become increasingly important as more complex graph-builder APIs are added.
+
+**In summary:**
+
+- array syntax: optimal but literally impossible
+- tuple syntax: magic but pretty
+- builder syntax: way too verbose for a pattern that's supposed to increase convenience
+- `vec!`-style macro: not any less magic than tuple syntax, not as pretty
+
+**Conclusion:** tuple syntax, powered by builder syntax under the hood that users can use instead when it's more convenient.

```

Comment:
The tuple syntax has an important limitation though: the number of systems it can take is limited, and the user can then get a confusing error while incrementally adding more systems to a set.

Reply:
Agreed. I'm currently on team macro as a result. I think that limitation is much more significant here than elsewhere.